### ISSUE TITLE
Add type() function to print type of variable

### ISSUE THREAD

**Issue by plajjan** [actonlang/acton#415]
There is no type() function. Type information is available at compile time but not readily available at run time. I suppose there is still some notion of type but how can we access it?

One can print types of things by asking actonc with --types but for folks coming from python or similar non-type / duck typed languages, it is normal to explore types at run time.

Can we have a type() function? This makes it easier for beginners to be productive with no or little overhead, right?

**Comment by plajjan** [comment]
Could potentially make this a function of the dev / release compiler profile if we think this is a bad idea to have in a final program?

**Comment by plajjan** [comment]
I just tried adding a line to the print function to also print the "type". Is class GCINFO the closest we have today? Maybe that is a good value, it just feels bad because it is about GC and not really the type, but maybe it's the same? I note we use it in other places to print the "type" of something.

```patch
diff --git a/builtin/builtin_functions.c b/builtin/builtin_functions.c
index 67967c4..2fae361 100644
--- a/builtin/builtin_functions.c
+++ b/builtin/builtin_functions.c
@@ -27,6 +27,7 @@ void $print(int size, ...) {
     va_start(args,size);
     if (size > 0) {
         $value elem = va_arg(args,$value);
+        printf("TYPE: %s\n", (const char*)elem->$class->$GCINFO);
         fputs((const char*)elem->$class->__str__(elem)->str,stdout);
     }
     for (int i=1; i<size; i++) {
```

and for an integer I get `TYPE: $int`... so we can use this. I have no clue how to extend the compiler though ;)

**Comment by nordlander** [comment]
It all depends on the contexts in which type() is supposed to be used. If type() really just returns compile-time type information one could imagine allowing things like `type(e)` inside type annotations, which would then mean "the type of expression `e`, whatever that type is". One could also assume a string representation of said compile-time type whenever `type(e)` is used in an expression context, as in `print(type(e))`.

But I'd like to raise a warning against the idea of letting `type(e)` denote _dynamic_ type information -- that is, type information only known at run-time. First of all it would require an apparatus for representing types as run-time terms; not only as simple class names but also including generic type parameters, plus a special representation of function and tuple types. Secondly, the pythonic idea that `type(e)` reveals the true type of `e` in every context would have disastrous effects on the safety of our type system, since one could no longer trust a generic function to be unable to scrutinize its generically typed parameters. For one thing, our scheme for preserving the integrity of the actors' states relies on this property.

In summary: `type(e)`as a type expression or as a string valued term expression that reveals only compile-time type information would be ok and rather straightforward to implement. Everything else should remain a no-no.

**Comment by plajjan** [comment]
I think that is okay, I just see this as a sort of stepping stone to gradually work users into a statically typed environment when they are coming from a dynamically typed one. I don't think most users really consider the dynamic part, it's just that they are not used to having a compiler but rather do print style debugging with type() and dir() of things to find out what they look like. The Actonic way to do this would be to inspect code or ask the compiler but if we can easily bridge a gap here by providing a type() function or maybe an object method or attribute? (If that is easier / better), I think that would be cool.

I wonder though, what sort of bad code people could write with this. I see it as a debug / learning kind of thing, not something to find permanently in code...

**Comment by plajjan** [comment]
I added this in #1587. Closing.
